compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -183,36 +176,23 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
         return null
     }
 
-    private fun gen(expression: IrElement, data: BlockInfo): StackValue = codegen.gen(expression, data)
-
-    private fun coerceNotToUnit(fromType: Type, fromKotlinType: KotlinType?, toKotlinType: KotlinType): StackValue =
-        codegen.coerceNotToUnit(fromType, fromKotlinType, toKotlinType)
-
     abstract inner class Switch(
         val subject: IrGetValue,
         val elseExpression: IrExpression?,
         val expressionToLabels: ArrayList<ExpressionToLabel>
     ) {
-        protected val endLabel = Label()
         protected val defaultLabel = Label()
 
         open fun shouldOptimize() = false
 
-        open fun genOptimizedIfEnoughCases(): StackValue? {
+        open fun genOptimizedIfEnoughCases(): PromisedValue? {
             if (!shouldOptimize())
                 return null
 
-            genSubject()
             genSwitch()
-            genThenExpressions()
-            val result = genElseExpression()
-
-            mv.mark(endLabel)
-            return result
+            return genBranchTargets()
```

Comment:
@ting-yuan Please review changes in SwitchGenerator.kt, as I understand there should be several more specific generators and method merging could break your plans/design
```
            genSubject()
            genSwitch()
            genThenExpressions()
```

Reply:
Although I prefer smaller functions as they are easier to read, this looks fine to me.